File: Plotter/Axis.m
//
//  Axis.m
//  Plotter
//
//  Created by Lawrence Kesteloot on 7/25/15.
//  Copyright (c) 2015 HeadCode. All rights reserved.
//

#import "Axis.h"

@interface Axis () {
    NSMutableArray *_seriesArray;
}

@end

@implementation Axis

@dynamic seriesArray;

- (id)init {
    self = [super init];
    
    if (self) {
	_minValue = 0;
	_maxValue = 0;
	_range = 0;
	_grid = nil;
	_seriesArray = [NSMutableArray array];
    }
    
    return self;
}

- (void)addSeries:(Series *)series {
    [_seriesArray addObject:series];
}

- (void)updateStats {
    BOOL first = YES;

    for (Series *series in _seriesArray) {
	if (first) {
	    _minValue = series.minValue;
	    _maxValue = series.maxValue;
	    first = NO;
	} else {
	    if (series.minValue < _minValue) {
		_minValue = series.minValue;
	    }
	    if (series.maxValue > _maxValue) {
		_maxValue = series.maxValue;
	    }
	}
    }

    _range = _maxValue - _minValue;

    _grid = [[Grid alloc] initForRangeWithMin:_minValue andMax:_maxValue];
}

- (NSArray *)seriesArray {
    return _seriesArray;
}

@end

File: src/core/axis.ts
// A left or right vertical axis.

import { Grid, makeRangeGrid } from "./grid.js";
import { type Series } from "./series.js";

export class Axis {
    // The series displayed by this axis.
    readonly seriesArray: Series[] = [];

    // The lowest and largest values we must display.
    minValue = 0;
    maxValue = 0;
    // Max minus min.
    range = 0;
    // The grid used for this axis. Created by updateStats().
    grid: Grid | undefined = undefined;

    addSeries(series: Series): void {
        this.seriesArray.push(series);
    }

    // Update the stats and create the grid after all series have been added.
    updateStats(): void {
        let first = true;

        for (const series of this.seriesArray) {
            if (first) {
                this.minValue = series.minValue;
                this.maxValue = series.maxValue;
                first = false;
            } else {
                this.minValue = Math.min(this.minValue, series.minValue);
                this.maxValue = Math.max(this.maxValue, series.maxValue);
            }
        }

        this.range = this.maxValue - this.minValue;
        this.grid = makeRangeGrid(this.minValue, this.maxValue);
    }
}
